chainer/links/connection/lstm.py

```diff
@@ -197,6 +211,11 @@ def __call__(self, x):
             ~chainer.Variable: Outputs of updated LSTM units.
 
         """
+        if self.upward.has_uninitialized_params:
+            in_size = x.size // len(x.data)
+            self.upward._initialize_params(in_size)
```

Comment:
I think I understand now. It seems that `Linear._initialize_params` only needs to be called from LSTM link because LSTM wants to use the performance optimization of grouping the 4 linear layers for each gate into a single matrix multiply and also needs to initialize the forget gate differently from the other gates. Otherwise, it would not be necessary to call `Linear._initialize_params` from the LSTM link. I can understand why it is written this way (to improve performance), but a slight drawback is that the initialization code gets trickier. 

Since `Linear._initialize_params` is only needed as a performance optimization, I think maybe it is better to keep this method private for now.

Perhaps another option to avoid calling this private method of Linear link would be to change the LSTM link so that it no longer uses Linear link at all and instead just calls the linear function directly.